File: cpp/course_schedule_II.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_set>
#include <map>
#include <algorithm>
#include <limits.h>
#include <math.h>

using namespace std;

/**
 * Definition for binary tree
 */
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

/**
 * Definition for singly-linked list.
 */
struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
 };

/**
 * Definition for undirected graph.
 * */
struct UndirectedGraphNode {
    int label;
    vector<UndirectedGraphNode *> neighbors;
    UndirectedGraphNode(int x) : label(x) {};
};

/**
 * Definition for binary tree with next pointer.
 */
struct TreeLinkNode {
  int val;
  TreeLinkNode *left, *right, *next;
  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
};

class Solution {
public:
	struct Node {
		unordered_set<int> neighbors;
	};

    vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
    	// [1, 0] 1 ---> 0
    	vector<Node> graphIncoming(numCourses);
    	vector<Node> graphOutgoing(numCourses);

    	for (auto p:prerequisites) {
    		graphIncoming[p.second].neighbors.insert(p.first);
    		graphOutgoing[p.first].neighbors.insert(p.second);
    	}

    	vector<int> sorted;
    	unordered_set<int> s;

    	for (int i=0; i<graphIncoming.size(); i++) {
    		if (graphIncoming[i].neighbors.size()==0) {
    			s.insert(i);
    		}
    	}

    	while (!s.empty()) {
    		int n=*s.begin();
    		sorted.push_back(n);
    		s.erase(s.begin());

    		for (auto it=graphOutgoing[n].neighbors.begin(); it!=graphOutgoing[n].neighbors.end(); it++) {
    			graphIncoming[*it].neighbors.erase(n);
    			if (graphIncoming[*it].neighbors.size()==0) {
    				s.insert(*it);
    			}
    		}
    		graphOutgoing[n].neighbors.clear();
    	}

    	int l=0, r=sorted.size()-1;
    	while (l<r) {
    		int temp=sorted[l];
    		sorted[l]=sorted[r];
    		sorted[r]=temp;
    		l++;
    		r--;
    	}

    	for (int i=0; i<graphIncoming.size(); i++) {
    		if (graphIncoming[i].neighbors.size()>0) {
    			vector<int> empty;
    			return empty;
    		}
    	}

    	return sorted;
    }
};

int main()
{
	return 0;
}

File: java/course_schedule_ii.java
public class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        if (numCourses==0) return new int[0];
        List<Integer> courses=new ArrayList<Integer>();
        HashSet<Integer>[] graph=new HashSet[numCourses];
        for (int i=0; i<numCourses; i++) {
            graph[i]=new HashSet<Integer>();
        }
        int[] incomingCnt=new int[numCourses];
        
        for (int i=0; i<prerequisites.length; i++) {
            if (!graph[prerequisites[i][1]].contains(prerequisites[i][0])) {
                graph[prerequisites[i][1]].add(prerequisites[i][0]);
                incomingCnt[prerequisites[i][0]]++;
            }
        }
        
        TreeSet<Integer> set=new TreeSet<Integer>();
        for (int i=0; i<incomingCnt.length; i++) {
            if (incomingCnt[i]==0) set.add(i);
        }
        
        while (!set.isEmpty()) {
            int n=set.pollFirst();
            courses.add(n);
            
            for (int neighbor:graph[n]) {
                incomingCnt[neighbor]--;
                if (incomingCnt[neighbor]==0) {
                    set.add(neighbor);
                }
            }
            graph[n].clear();            
        }
        
        for (int i=0; i<graph.length; i++) {
            if (!graph[i].isEmpty()) return new int[0];
        }
        
        int[] res=new int[courses.size()];
        for (int i=0; i<courses.size(); i++) {
            res[i]=courses.get(i);
        }
        return res;
    }
}
